### ISSUE TITLE
OS X: @try/@catch test failing on node v0.12.0

### ISSUE THREAD

**Issue by TooTallNate** [node-ffi/node-ffi#195]
/cc @kjunichi @LinusU

__From @LinusU__: That one test is failing for me with `node v0.12.0` installed via Homebrew on MacBook Air 6,2.

```sh
git clone git@github.com:node-ffi/node-ffi.git
cd node-ffi
npm install
npm test
```

```text
  1 failing

  1) @try / @catch should throw a Buffer instance when an exception happens:
     AssertionError: false == true
      at Context.<anonymous> (/private/tmp/node-ffi/test/objc.js:47:9)
      at callFn (/private/tmp/node-ffi/node_modules/mocha/lib/runnable.js:266:21)
      at Test.Runnable.run (/private/tmp/node-ffi/node_modules/mocha/lib/runnable.js:259:7)
      at Runner.runTest (/private/tmp/node-ffi/node_modules/mocha/lib/runner.js:387:10)
      at /private/tmp/node-ffi/node_modules/mocha/lib/runner.js:470:12
      at next (/private/tmp/node-ffi/node_modules/mocha/lib/runner.js:312:14)
      at /private/tmp/node-ffi/node_modules/mocha/lib/runner.js:322:7
      at next (/private/tmp/node-ffi/node_modules/mocha/lib/runner.js:257:23)
      at Immediate._onImmediate (/private/tmp/node-ffi/node_modules/mocha/lib/runner.js:289:5)
      at processImmediate [as _immediateCallback] (timers.js:358:17)
```

It seems to me that C++ exceptions are disabled at runtime for some reason on v0.12.0 of node.

**Comment by mroderick** [comment]
With latest master (08c91c29005ee6c33cd01f3936c1d7960ce2f1f3), I am seeing two test failures in node `0.12.6` installed via nvm 

```bash
  1) @try / @catch should proxy @try/@catch to JavaScript via try/catch/throw:
     AssertionError: Missing expected exception..
      at Function._throws (assert.js:290:5)
      at Function.assert.throws (assert.js:307:11)
      at Context.<anonymous> (test/objc.js:33:14)

  2) @try / @catch should throw a Buffer instance when an exception happens:
     AssertionError: Missing expected exception..
      at Function._throws (assert.js:290:5)
      at Function.assert.throws (assert.js:307:11)
      at Context.<anonymous> (test/objc.js:40:14)
```

**Comment by TooTallNate** [comment]
Right, both tests are failing indeed. In https://github.com/node-ffi/node-ffi/commit/a93d5a350076386bc093fdc9c2d86c15c38bb3d0 I made the tests more explicit. I'm still not sure of what change in Node is causing this.

**Comment by TooTallNate** [comment]
@indutny You wouldn't happen to know the culprit here would you? Basically at some point in between v0.10 and v0.12 Node.js, C++ exceptions can no longer be enabled in native addons.

**Comment by indutny** [comment]
@TooTallNate we have them disabled in node.js, so it may be the case?

**Comment by TooTallNate** [comment]
@indutny Right, but in node-ffi we're (attempting to) re-enable them in the binding.gyp file (https://github.com/node-ffi/node-ffi/blob/f954bb7464a905a1e11101d68547a193337fbf43/binding.gyp#L24). This worked though v0.10, but something during v0.11-v0.12 made this no longer work. I don't want you to spend too much time here, just wondering if you knew off the top of your head what's the culprit.

**Comment by indutny** [comment]
@TooTallNate sorry, but nothing pops in immediately. Are you sure that enabling this in `binding.gyp` is enough? I don't think that we are linking node with the runtime which is required for support exceptions. cc @bnoordhuis

**Comment by TooTallNate** [comment]
All I know is that nothing changed in this repo in between the v0.10 working and v0.12 failing. I've tried to `git bisect` but nan doesn't support v0.11 patch versions so I makes it a bit harder.

I haven't tried to make a narrowed down case yet though...

**Comment by kjunichi** [comment]
Hi,it's been a long time.

We need 

```c
dlopen("/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa",RTLD_NOW);
```

when we use v4.1.1(maybe above v0.12)

So, just change to test code following,then we get success all of tests in osx.

First,　we make a small library to dlopen.

```c
#include <dlfcn.h>

void foo() {
        dlopen("/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa",RTLD_NOW);
}
```

```bash
gcc -dynamiclib -undefined suppress -flat_namespace mylib.c -o libmy.dylib
```

and put it test folder.

Next, we add this code to objc.js

```diff
  afterEach(gc);
+ var dummyLib = new ffi.Library('./test/libmy.dylib',{'foo':['void',[]]});
+ dummyLib.foo();
  var objcLib = new ffi.Library('libobjc', {
```

I think #211 is related issues.

Thank you.

**Comment by bnoordhuis** [comment]
> C++ exceptions can no longer be enabled in native addons.

You can use C++ exceptions in add-ons but there should never be any `-fno-exceptions` stack frames between the thrower and the catcher.  That includes JS stack frames; throwing a C++ exception from a JS -> C++ callback is unsafe.

**Comment by TooTallNate** [comment]
Thanks for the hint @kjunichi. It looks like you're right in that we have to load Cocoa before Objective-C exceptions work (perhaps something with loading `NSException` into memory?). Anyways, not entirely sure _why_ that fixes it, but fixed the two failing tests in c81ab1e.
